[SYCL][Graph] Support for sycl_ext_oneapi_enqueue_barrier extension

mfrancepillois opened the pull request:
Adds support to handle barrier enqueuing with Record&Replay API. Barriers are implemented as empty nodes enforcing the required dependencies.

Adds tests that check:
1) correctness of graph structure when barriers have been enqueued, 
2) processing behavior, 
3) exception throwing if barriers are used within explicit API.

Notes:
1) Barriers can only be used with Record&Replay API, since barriers rely on events to enforce dependencies.
2) The scope of a barrier is limited to the Graph in which the barrier was added. As a result, if a barrier with no explicit events to wait for is part of a graph, when the graph is executed this barrier waits for all the kernels/operations in the graph that were previously submitted. However, the barrier does not wait for kernels/operations outside the graph that might have been submitted to the same execution queue.



---------

Bensuo commented on sycl/source/handler.cpp:
```diff
@@ -368,11 +368,28 @@ event handler::finalize() {
-                                             MCodeLoc));
+  case detail::CG::BarrierWaitlist: {
+    if (auto GraphImpl = getCommandGraph(); GraphImpl != nullptr) {
+      // if no event to wait for was specified, we add all the previous
+      // nodes/events of the graph
```
```suggestion
      // if no event to wait for was specified, we add all exit
      // nodes/events of the graph
```
Small clarification since we're not adding all nodes here but only the exit/leaf nodes.

EwanC commented:
Hi @intel/llvm-reviewers-runtime, could we have a review of this please

againull commented on sycl/unittests/Extensions/CommandGraph.cpp:
```diff
@@ -1306,10 +1219,185 @@ TEST_F(CommandGraphTest, InOrderQueueWithEmptyLast) {
+  //     (B2)
+  //     /|\ 
+  //    / | \
+  // (6) (7) (8) (those nodes also have as a precedessor)
```
NIT: those nodes also have _barrier_ as a predecessor?

mfrancepillois replied:
Thanks. The sentence should have been "those nodes also have B1 as a predecessor".
We fixed it.

al42and commented:
> Multi-queues barrier is not supported

Does it mean that `q.ext_oneapi_submit_barrier({e1, e2});` is forbidden if `e1` and `e2` come from different queues?

Or does it forbid any mixing of the queues, including the following code?

```cpp
sycl::event e = q1.ext_oneapi_submit_barrier();
q2.ext_oneapi_submit_barrier({e});
```

mfrancepillois commented:
> > Multi-queues barrier is not supported
> 
> Does it mean that `q.ext_oneapi_submit_barrier({e1, e2});` is forbidden if `e1` and `e2` come from different queues?
> 
> Or does it forbid any mixing of the queues, including the following code?
> 
> ```c++
> sycl::event e = q1.ext_oneapi_submit_barrier();
> q2.ext_oneapi_submit_barrier({e});
> ```
The sentence was not accurate. Indeed, the implementation supports both cases you mentioned. We added tests (unitests) to verify these use cases. We also updated the PR description to better explain the limitations of our implementation.

mfrancepillois commented:
@intel/llvm-gatekeepers could you please merge this PR?

steffenlarsen commented on sycl/unittests/Extensions/CommandGraph.cpp:
```diff
@@ -1306,10 +1219,282 @@ TEST_F(CommandGraphTest, InOrderQueueWithEmptyLast) {
+  //   \  |  /
+  //    \ | /
+  //     (B)
+  //     / \
```
It looks like post-commit doesn't like these `\` despite them being in comments. @mfrancepillois - Could you please address this issue ASAP?

EwanC replied:
Turns out it's the trailing whitespace character that triggers the warning https://github.com/intel/llvm/pull/11493
Can see it show in godbolt too https://godbolt.org/z/veeYvs6de if you add a whitespace character to the end of a line
